### PR TITLE
Olivierg/dogstatsd as fxapp

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?

Converts the DogstatsD binary to an Fx App.

### Motivation

Top-down approach to componentization.

### Describe how to test/QA your changes

Run the DogStatsD `start` commands  and verify that command-line flags still work. 
Also perform the following checks:
- Check DogStatsD starts on Windows.
- Check the Agent starts with the default configuration path (without using `-c`)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: cmd/dogstatsd/main.go [hunk 1]
```diff
@@ -52,14 +71,33 @@ on dashboards. DogStatsD implements the StatsD protocol, along with a few
...
 		Short: "Start DogStatsD",
 		Long:  `Runs DogStatsD in the foreground`,
-		RunE:  start,
+		RunE: func(*cobra.Command, []string) error {
+			return fxutil.OneShot(start,
+				fx.Supply(cliParams),
+				fx.Supply(core.BundleParams{
+					ConfFilePath:      cliParams.confPath,
+					ConfigLoadSecrets: false,
+					ConfigMissingOK:   true,
```
Comment (ogaca-dd) [inline cmd/dogstatsd/main.go:93]: @djmitche : Are these previous value correct (`ConfigLoadSecrets`, `ConfigMissingOK`)?

Comment (djmitche) [reply cmd/dogstatsd/main.go:93]: That's a good question.  You'd need to look at the implementation of `runAgent`, below, and `pkgconfig.Load()`, to see how it behaved before.  Note that you may also need to change `core.BundleParams#ConfigName`.

If there's other required functionality that's missing in comp/core/config, please feel invited to add it!

File: pkg/util/fxutil/error_handler.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+	"os"
+
+	"go.uber.org/fx"
+)
+
+type errorHandler struct{}
+
+func (errorHandler) HandleError(error) {
+	if os.Getenv("TRACE_FX") == "" {
```
Comment (djmitche) [inline pkg/util/fxutil/error_handler.go:18]: did you want to use `traceFx` here?

**Review by djmitche — APPROVED** [review]
This is looking good!  I have a few additional points, but they're fine as follow-ups rather than changes in this PR.

 * It'd be great to have `cmd/dogstatsd/command.MakeCommands` and the subcommands defined in `cmd/dogstatsd/subcommands`, so that `cmd/dogstatsd` itself has very little code in it -- to match `cmd/agent`.  That's just moving some functions around, really.
 * `dogstatsd version` could share an implementation with `agent version` and `dogstatsd-cluster-agent version` -- @juliogreff is working on a similar thing in #13967.
 * The `--cfgpath` flag is handled differently here than in the main agent: the Cobra flag is attached to the config values directly.  Could those be harmonized without changing user-visible behavior?
 * The windows service startup code in `cmd/dogstatsd/main_windows.go` is duplicative of that in `cmd/agent/windows/service`.  How could we share implementations there?  This might be something to solve when we start running daemons using `fxutil.Run` instead of `fxutil.OneShot`.  Maybe `dogstatsd` also grows `start-service` and `stop-service` subcommands on Windows, and a `stop` subcommand on linux?

File: comp/core/internal/params.go [hunk 3]
```diff
@@ -48,6 +48,14 @@ type BundleParams struct {
 	// LoggerName is the name that appears in the logfile
 	LoggerName string
 
+	// ExcludeDefaultConfPath determines whether the default config path must be excluded.
+	ExcludeDefaultConfPath bool
+
+	// DefaultConfPath determines the default configuration path.
+	// if DefaultConfPath is empty, then the default configuration path of the Agent is used.
+	// if ExcludeDefaultConfPath is true this field is ignored.
+	DefaultConfPath string
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply comp/core/internal/params.go:57]: I added this field to not change the current behavior of DogStatsD and had in mind to have a follow up PR to cleanup default configuration paths for all Agents (DogStatsd binary defines a default configuration path on Windows but not on Linux. I guess there are some differences between the Core Agent and the other agents regarding default configuration and I am not sure this is expected). What do you think?

File: tasks/dogstatsd.py [hunk 4]
```diff
@@ -27,7 +27,7 @@
 # constants
 DOGSTATSD_BIN_PATH = os.path.join(".", "bin", "dogstatsd")
 STATIC_BIN_PATH = os.path.join(".", "bin", "static")
-MAX_BINARY_SIZE = 32 * 1024
+MAX_BINARY_SIZE = 35 * 1024
```
Comment (mx-psi) [inline tasks/dogstatsd.py:30]: Is this change related to the PR?

Comment (ogaca-dd) [reply tasks/dogstatsd.py:30]: Yes, as the size of `dogstatsd` has increased.

**Review by mx-psi — APPROVED** [review]

File: comp/core/internal/params.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (djmitche) [reply comp/core/internal/params.go:57]: Doing so in a followup sounds good.  I think we should aim to make things "the same" where possible, and where that's not possible to make the differences clear.

File: comp/core/internal/params.go [hunk 5]
```diff
@@ -48,6 +48,14 @@ type BundleParams struct {
 	// LoggerName is the name that appears in the logfile
 	LoggerName string
 
+	// ExcludeDefaultConfPath determines whether the default config path must be excluded.
```
Comment (remeh) [inline comp/core/internal/params.go:51]: Is this exclusive to DogStatsD? It doesn't seem to be used by the Agent in the component config. I'm not sure it's worth sharing it as a global bundle parameter if I'm right.

**Review by remeh — COMMENTED** [review]
LGTM not much to say

I hope `BundleParams` won't grow too much with every app or it can easily become verbose...

File: comp/core/internal/params.go (hunk 5 above)
Comment (ogaca-dd) [reply comp/core/internal/params.go:51]: I have tried to not update the behavior of dogstatsd in this PR but there will be a followup PR to update this code: See https://github.com/DataDog/datadog-agent/pull/13842#discussion_r1001605760.

**Review by remeh — APPROVED** [review]
